![](https://media.giphy.com/media/JWuBH9rCO2uZuHBFpm/giphy.gif)

Remember the rules of ~Fight~ Code Club:
1. ALWAYS DOCUMENT
2. Cite resources that you use (paste links)
3. Include the names people who you worked with
4. Be neat and organized

### Question 1

Use the [pokeAPI](https://pokeapi.co/)

Find out information about the Pokemon "Gible":
- What is its type(s)?
- What is its abilities?
- How many pokemon are "dragon" type?

You must use python and apply the GET method via their API to display this information.

##### What is its type(s)?

In [67]:
import requests #importing package requests
gible_url = ('https://pokeapi.co/api/v2/pokemon/gible') 
#defining a url to find information on the Pokemon Gible as "gible_url"

response_type = requests.get(gible_url) 
#defining the request as response_type
#using the library called requests that was imported, then use .get() function to make a GET request to the gibel_url 

print(response_type) #print response for finding the types in the next cell                                       

<Response [200]>


In [56]:
gible = response_type.json() 
#parse response of the GET request into JSON format to later view as raw JSON data and save that JSON info as "gible_type" 

pp = pprint.PrettyPrinter(indent=1) 
#defining parameters as pp, then use PrettyPrinter instance to understand the data by 1 indent parameter
                                    
pp.pprint(gible) #print gible data returned from the GET request as JSON with pprint function

{'abilities': [{'ability': {'name': 'sand-veil',
                            'url': 'https://pokeapi.co/api/v2/ability/8/'},
                'is_hidden': False,
                'slot': 1},
               {'ability': {'name': 'rough-skin',
                            'url': 'https://pokeapi.co/api/v2/ability/24/'},
                'is_hidden': True,
                'slot': 3}],
 'base_experience': 60,
 'forms': [{'name': 'gible',
            'url': 'https://pokeapi.co/api/v2/pokemon-form/443/'}],
 'game_indices': [{'game_index': 443,
                   'version': {'name': 'diamond',
                               'url': 'https://pokeapi.co/api/v2/version/12/'}},
                  {'game_index': 443,
                   'version': {'name': 'pearl',
                               'url': 'https://pokeapi.co/api/v2/version/13/'}},
                  {'game_index': 443,
                   'version': {'name': 'platinum',
                               'url': 'https://pokeapi.co/api/v2/version/1

In [59]:
gible['types']#show the list of types (nested dictionary) of the JSON data called gible
#It maps the dictionary key "type" to another dictionary that maps the name and the url to values


[{'slot': 1,
  'type': {'name': 'dragon', 'url': 'https://pokeapi.co/api/v2/type/16/'}},
 {'slot': 2,
  'type': {'name': 'ground', 'url': 'https://pokeapi.co/api/v2/type/5/'}}]

In [58]:
#gible['types'] is a list of types dictionaries
#in each type dictionary of gible['types'] is a nested dictionary. It maps the dictionary key "type" to another dictionary that maps the name and the url to values

for types_dict in gible['types']: #for each types dictionary in the list gible['types']
    types = types_dict['type'] # to access the nested dictionary with name info, I am first defining the key 'types' as a key within a types_dict
    types_name = types['name'] #I am defining the key "types_name" as a name key within the dictionary types
    print("Gible type:",types_name)

Gible type: dragon
Gible type: ground


##### What is its abilities?

In [60]:
gible['abilities'] #of the JSON format data called gible, show the abilities
#a list of ability (nested) dictionaries

[{'ability': {'name': 'sand-veil',
   'url': 'https://pokeapi.co/api/v2/ability/8/'},
  'is_hidden': False,
  'slot': 1},
 {'ability': {'name': 'rough-skin',
   'url': 'https://pokeapi.co/api/v2/ability/24/'},
  'is_hidden': True,
  'slot': 3}]

In [61]:
for abilities_dict in gible['abilities']: #for each abilities dictionary in the list gible['abilities']
    abilities = abilities_dict['ability'] # to access the nested dictionary with name info, I am first defining the key 'abilities' as a key within a abilities_dict
    abilities_name = abilities['name'] #I am defining the key "abilities_name" as a name key within the dictionary abilties
    print("Gible ability:",abilities_name)

Gible ability: sand-veil
Gible ability: rough-skin


##### How many pokemon are "dragon" type?

In [68]:
#Dragon type
dragon_url = ('https://pokeapi.co/api/v2/type/16/') 
#defining a url to find information on Pokemon that are dragon type as "dragon_url"
                                                   
response_dragon = requests.get(dragon_url) 
#defining the request as response_dragon, using the library called requests that was imported, then use .get() function to make a GET request to the dragon_url 

print(response_dragon) #print response for finding the dragon types in the next cell                                       

<Response [200]>


In [69]:
dragon = response_dragon.json() 
#parse response of the GET request into JSON format and save that JSON info as "dragon" 

pp = pprint.PrettyPrinter(indent=1)
#defining parameters as pp, then use PrettyPrinter instance to understand the data by 1 indent parameter
 
pp.pprint(dragon) #printing nicely with pprint for the dragon data returned from the GET request

{'damage_relations': {'double_damage_from': [{'name': 'ice',
                                              'url': 'https://pokeapi.co/api/v2/type/15/'},
                                             {'name': 'dragon',
                                              'url': 'https://pokeapi.co/api/v2/type/16/'},
                                             {'name': 'fairy',
                                              'url': 'https://pokeapi.co/api/v2/type/18/'}],
                      'double_damage_to': [{'name': 'dragon',
                                            'url': 'https://pokeapi.co/api/v2/type/16/'}],
                      'half_damage_from': [{'name': 'fire',
                                            'url': 'https://pokeapi.co/api/v2/type/10/'},
                                           {'name': 'water',
                                            'url': 'https://pokeapi.co/api/v2/type/11/'},
                                           {'name': 'grass',
                    

In [70]:
dragon['pokemon'] #a list of pokemon dictionaries of the JSON format data called dragon

[{'pokemon': {'name': 'dratini',
   'url': 'https://pokeapi.co/api/v2/pokemon/147/'},
  'slot': 1},
 {'pokemon': {'name': 'dragonair',
   'url': 'https://pokeapi.co/api/v2/pokemon/148/'},
  'slot': 1},
 {'pokemon': {'name': 'dragonite',
   'url': 'https://pokeapi.co/api/v2/pokemon/149/'},
  'slot': 1},
 {'pokemon': {'name': 'kingdra',
   'url': 'https://pokeapi.co/api/v2/pokemon/230/'},
  'slot': 2},
 {'pokemon': {'name': 'vibrava',
   'url': 'https://pokeapi.co/api/v2/pokemon/329/'},
  'slot': 2},
 {'pokemon': {'name': 'flygon',
   'url': 'https://pokeapi.co/api/v2/pokemon/330/'},
  'slot': 2},
 {'pokemon': {'name': 'altaria',
   'url': 'https://pokeapi.co/api/v2/pokemon/334/'},
  'slot': 1},
 {'pokemon': {'name': 'bagon',
   'url': 'https://pokeapi.co/api/v2/pokemon/371/'},
  'slot': 1},
 {'pokemon': {'name': 'shelgon',
   'url': 'https://pokeapi.co/api/v2/pokemon/372/'},
  'slot': 1},
 {'pokemon': {'name': 'salamence',
   'url': 'https://pokeapi.co/api/v2/pokemon/373/'},
  'slot': 1

In [72]:
print( "Number of Dragon Pokemon:",len(dragon['pokemon'])) 
#printing the number of pokemon from the dragon data, using len function to tell me how many objects are ther

for pokemon_dict in dragon['pokemon']: #for each pokemon dictionary in the list dragon['pokemon']
    pokemon = pokemon_dict['pokemon'] #defining the key pokemon as a key within a pokemon_dict o access the nested dictionary with name info
    pokemon_name = pokemon['name'] #defining the key pokemon_name as a name key within the dictionary pokemon
    print(pokemon_name) #printing the pokemon_name

Number of Dragon Pokemon: 78
dratini
dragonair
dragonite
kingdra
vibrava
flygon
altaria
bagon
shelgon
salamence
latias
latios
rayquaza
gible
gabite
garchomp
dialga
palkia
giratina-altered
axew
fraxure
haxorus
druddigon
deino
zweilous
hydreigon
reshiram
zekrom
kyurem
dragalge
tyrunt
tyrantrum
goomy
sliggoo
goodra
noibat
noivern
zygarde
turtonator
drampa
jangmo-o
hakamo-o
kommo-o
guzzlord
naganadel
applin
flapple
appletun
dracozolt
dracovish
duraludon
dreepy
drakloak
dragapult
eternatus
regidrago
giratina-origin
kyurem-black
kyurem-white
charizard-mega-x
ampharos-mega
garchomp-mega
latias-mega
latios-mega
sceptile-mega
altaria-mega
rayquaza-mega
salamence-mega
exeggutor-alola
zygarde-10
zygarde-50
zygarde-complete
kommo-o-totem
necrozma-ultra
flapple-gmax
appletun-gmax
duraludon-gmax
eternatus-eternamax


### Question 2

Use the [Open Air Quality API](https://docs.openaq.org/)

Hint: typically queries need a ? in the url

Find out how much ozone there is near CCNY. Use the most recent date that you pull the information.

You mut use python and apply the GET method via their API and display this information.

In [86]:
import requests #importing package requests

ccny_ozone_url= ('https://api.openaq.org/v1/measurements?'
                 'city=New+York-Northern+New+Jersey-Long+Island'
                 '&location=CCNY'
                 '&parameter=o3')
#defining a url to find information for CCNY ozone as ccny_ozone_url

response_ccny = requests.get(ccny_ozone_url) 
#defining the request as response_ccny
#using the library called requests that was imported, then use .get() function to make a GET request to the ccny_ozone_url

response_ccny

ccny_ozone=response_ccny.json() 
#parse response of the GET request into JSON format and save that JSON info as ccny_ozone 

pp = pprint.PrettyPrinter(indent=1)
#defining parameters as pp, then use PrettyPrinter instance to understand the data by 1 indent parameter
 
pp.pprint(ccny_ozone)
#print CCNY ozone data returned from the GET request as JSON with pprint function

{'meta': {'found': 65584,
          'license': 'CC BY 4.0d',
          'limit': 100,
          'name': 'openaq-api',
          'page': 1,
          'website': 'https://u50g7n0cbj.execute-api.us-east-1.amazonaws.com/'},
 'results': [{'city': 'New York-Northern New Jersey-Long Island',
              'coordinates': {'latitude': 40.8197, 'longitude': -73.9481},
              'country': 'US',
              'date': {'local': '2021-10-27T08:00:00-04:00',
                       'utc': '2021-10-27T12:00:00Z'},
              'location': 'CCNY',
              'parameter': 'o3',
              'unit': 'ppm',
              'value': 0.026},
             {'city': 'New York-Northern New Jersey-Long Island',
              'coordinates': {'latitude': 40.8197, 'longitude': -73.9481},
              'country': 'US',
              'date': {'local': '2021-10-27T07:00:00-04:00',
                       'utc': '2021-10-27T11:00:00Z'},
              'location': 'CCNY',
              'parameter': 'o3',
           

##### Find out how much ozone there is near CCNY. Use the most recent date that you pull the information.


In [89]:
print(ccny_ozone['results'][0]) 
#printing result from the dictionary of ccny_ozone, [0] is the index for the first which is most recent result from that list 

{'location': 'CCNY', 'parameter': 'o3', 'value': 0.026, 'date': {'utc': '2021-10-27T12:00:00Z', 'local': '2021-10-27T08:00:00-04:00'}, 'unit': 'ppm', 'coordinates': {'latitude': 40.8197, 'longitude': -73.9481}, 'country': 'US', 'city': 'New York-Northern New Jersey-Long Island'}


### Question 4

Use the [Open Air Quality API](https://docs.openaq.org/)

Hint: typically queries need a ? in the url. Note that cities that need a space use a %20 between them (ie Los%20Angeles). Make sure that when you request information about a measurement that you are using a url that can do this for you.

First find out what the parameter id is of 'Particulate matter less than 10 micrometers in diameter.'

Then, find out how much 'Particulate matter less than 10 micrometers in diameter' there is in:
- Beijing
- Addis Ababa

Bonus points if you can aggregate this information in one table.

You must use python and apply the GET method via their API and display this information.



In [99]:
#creating a Query URL and defining it as measurement_url

measurement_url = ('https://docs.openaq.org/v2/measurements?'
          'city[]=Beijing'
          '&city[]=Addis%Ababa'
          '&parameter=pm10')
#defining a url to find information for Particulate matter less than 10 micrometers in diameter in Beijing and Addis Ababa as measurement_url

response = requests.get(measurement_url) 
#defining the request as response
#using the library called requests that was imported, then use .get() function to make a GET request to the measurement_url

response #print response for finding the information of particulate matter in the areas

measurement=response.json() 
#parse response of the GET request into JSON format to later view as raw JSON data and save that JSON info as measurement

pp = pprint.PrettyPrinter(indent=1)
#defining parameters as pp, then use PrettyPrinter instance to understand the data by 1 indent parameter

pp.pprint(measurement)
#print particulate matter data returned from the GET request as JSON with pprint function

{'meta': {'found': 6163153624,
          'license': 'CC BY 4.0d',
          'limit': 100,
          'name': 'openaq-api',
          'page': 1,
          'website': 'https://u50g7n0cbj.execute-api.us-east-1.amazonaws.com/'},
 'results': [{'city': None,
              'coordinates': {'latitude': 19.359226, 'longitude': -99.2047},
              'country': 'MX',
              'date': {'local': '2021-10-27T21:15:00-05:00',
                       'utc': '2021-10-28T02:15:00+00:00'},
              'entity': 'community',
              'isAnalysis': False,
              'isMobile': False,
              'location': 'LAS AGUILAS',
              'locationId': 64394,
              'parameter': 'pm10',
              'sensorType': 'low-cost sensor',
              'unit': 'µg/m³',
              'value': 17.3},
             {'city': None,
              'coordinates': {'latitude': 37.257515, 'longitude': -122.017265},
              'country': 'US',
              'date': {'local': '2021-10-27T19:15:00-07:

In [100]:
type(measurement['results']) #getting the type to verify that it is a list, then convert it into dataframe

list

In [101]:
import pandas as pd #importing the package pandas as pd
df = pd.DataFrame(measurement['results'])  #defining dataframe and add the results of measurment information, then putting them into the dataframe
df #printing dataframe of measurament results

locationId                            location parameter  value  \
0        64394                         LAS AGUILAS      pm10   17.3   
1        62983                          JZ Outside      pm10    2.2   
2       233478                        22546 Summit      pm10    0.8   
3       226722                          P_Air_M001      pm10    2.2   
4        73716            Pearlgrass Ct. San Ramon      pm10    0.7   
..         ...                                 ...       ...    ...   
95       67047                       Hyde Park, VT      pm10    2.8   
96       70183                        Outdoor Ynez      pm10    1.0   
97       62213  Clean Air Carolina Monkey Junction      pm10   13.2   
98       71046                             Laidley      pm10    6.2   
99       70525             Flowery Branch, Georgia      pm10   11.6   

                                                 date   unit  \
0   {'utc': '2021-10-28T02:15:00+00:00', 'local': ...  µg/m³   
1   {'utc': '2021-10-28T02:15:00+00:00', 'local': ...  µg/m³   
2   {'utc': '2021-10-28T02:15:00+00:00', 'local': ...  µg/m³   
3   {'utc': '2021-10-28T02:15:00+00:00', 'local': ...  µg/m³   
4   {'utc': '2021-10-28T02:15:00+00:00', 'local': ...  µg/m³   
..                                                ...    ...   
95  {'utc': '2021-10-28T02:15:00+00:00', 'local': ...  µg/m³   
96  {'utc': '2021-10-28T02:15:00+00:00', 'local': ...  µg/m³   
97  {'utc': '2021-10-28T02:15:00+00:00', 'local': ...  µg/m³   
98  {'utc': '2021-10-28T02:15:00+00:00', 'local': ...  µg/m³   
99  {'utc': '2021-10-28T02:15:00+00:00', 'local': ...  µg/m³   

                                          coordinates country  city  isMobile  \
0      {'latitude': 19.359226, 'longitude': -99.2047}      MX  None     False   
1   {'latitude': 37.257515, 'longitude': -122.017265}      US  None     False   
2    {'latitude': 37.142136, 'longitude': -121.97597}      US  None     False   
3    {'latitude': 38.511448, 'longitude': -121.54559}      US  None     False   
4    {'latitude': 37.787605, 'longitude': -121.92728}      US  None     False   
..                                                ...     ...   ...       ...   
95    {'latitude': 44.602146, 'longitude': -72.62728}      US  None     False   
96  {'latitude': 34.067856, 'longitude': -118.129295}      US  None     False   
97    {'latitude': 34.148354, 'longitude': -77.88137}      US  None     False   
98   {'latitude': 37.73978, 'longitude': -122.429565}      US  None     False   
99     {'latitude': 34.17047, 'longitude': -83.92868}      US  None     False   

   isAnalysis     entity       sensorType  
0       False  community  low-cost sensor  
1       False  community  low-cost sensor  
2        None  community  low-cost sensor  
3        None  community  low-cost sensor  
4       False  community  low-cost sensor  
..        ...        ...              ...  
95      False  community  low-cost sensor  
96      False  community  low-cost sensor  
97      False  community  low-cost sensor  
98      False  community  low-cost sensor  
99      False  community  low-cost sensor  

[100 rows x 13 columns]

# Question 5


Using Beautiful soup, get the subscriber number from this website: https://socialblade.com/youtube/user/pewdiepie/realtime 

In [122]:
from bs4 import BeautifulSoup #importing the package, BeautifulSoup from beautifulsoup4

twitch_url = 'https://twitchtracker.com/thatsrb2dude/subscribers'
#defining a url to find information for subscriber as twitch_url

response_twitch = requests.get(('https://twitchtracker.com/thatsrb2dude/subscribers'), headers={'User-Agent':'Chrome'}) 
#defining the request as response_twitch
#using the library called requests that was imported, then use .get() function to make a GET request to the url, 'https://twitchtracker.com/thatsrb2dude/subscribers', then choose the header of User-Agent from Chrome

                                   
response_twitch #printing response for the information on subscriber from the twitch user

<Response [200]>


In [136]:
twitch_soup = BeautifulSoup(response_twitch.text, 'html.parser') 
#defining the new key twich_soup, Using Requests to scrape data for Beautiful Soup to pull data out of the website

print(twitch_soup.prettify()) 
#printing the data using prettify() to turn a BeautifulSoup parse tree into a nicely formatted unicode string

#SOURCE: https://stackoverflow.com/questions/52389692/beautifulsoup-and-prettify-function

<!DOCTYPE html>
<html lang="en">
 <head>
  <meta charset="utf-8"/>
  <meta content="IE=edge" http-equiv="X-UA-Compatible"/>
  <meta content="width=device-width, initial-scale=1.0" name="viewport"/>
  <title>
   ThatSrb2DUDE's Subs Count and Statistics · TwitchTracker
  </title>
  <meta content="ThatSrb2DUDE detailed subscribers count grouped by month and sub tier" name="description"/>
  <meta content="twitch, statistics, stream, channel, stats, games, subs, subscribers, clips, viewers, followers, views, shroud, ninja, drdisrespect, summit1g, lirik, dreadztv" name="keywords"/>
  <script async="" crossorigin="anonymous" src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?client=ca-pub-3581129455196386">
  </script>
  <script>
   window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
        ga('create', 'UA-51794509-3', 'auto'); ga('send', 'pageview');
  </script>
  <script async="" src="https://www.google-analytics.com/analytics.js">
  </script>

In [139]:
twitch_soup.find('span', {'class':'to-number'}) 
#using beautiful soup to find the subscriber number form the website

<span class="to-number">174</span>

##### The number of suscriber the twitch account is 174 